### DISCUSSION TITLE
CSS modules are not ready for production

### DISCUSSION THREAD

**Discussion by 10xchs** [vercel/next.js#15819] [Help]
Using CSS modules, pages are incredibly ugly for a second before the CSS modules are sent down to the client.

I really, really, really don't want to use styled jsx anymore, or create global css files for that matter.

Nowhere in the docs does it say that CSS modules are second-rate in this fashion. There should be a huge disclaimer. Honestly they shouldn't even have been included when they're clearly not ready for production.

**Comment by nstfkc** [comment]
I am also having this problem with tailwindcss, do you think might be related with this?

**Comment by awareness481** [comment]
Update your **Next.js** version to 9.5.
There was a known issue with FOUCs (flash-of-unstyled-content) but it was fixed in the 9.5 release.

**Reply by 10xchs** [reply]
Is this true? They make no mention of it in the changelog.

**Reply by awareness481** [reply]
Yep, see this [comment](https://github.com/vercel/next.js/issues/13058#issuecomment-666948357)    
They also mention it in the [releases](https://github.com/vercel/next.js/releases) (CTRL+F fouc) but I think they don't list bugfixes on the nextjs.org announcements.

**Reply by 10xchs** [reply]
Updating from 9.4.4 to 9.5.1 did not fix my problem. But I'm glad it shouldn't behave like this and something is wrong. Also, people seem to be reporting only have problems in development but for me it's in production as well.

**Reply by aychtang** [reply]
I have same issue with production FOUC on 9.5.x

**Comment by Timer** [comment]
Next.js has extensive integration tests that verifies your app doesn't FOUC on initial load.

If you are experiencing a FOUC, there'd be a few reasons:

1. You have a non-standard `next.config.js` overriding our webpack config
2. You have a `pages/_document.js` (or other ext) that incorrectly sets styles
3. You're using a CSS-in-JS library without appropriate integration.

Could you please post your `pages/_document.js` and `next.config.js` for us to take a look? Thanks!

**Reply by 10xchs** [reply]
Oh. Sure, thanks!

`next.config.js`:

```
const path = require('path')

module.exports = {
  webpack: config => {
    config.resolve.alias['@'] = path.resolve(__dirname)
    return config
  },
  sassOptions: {
    includePaths: [path.resolve(__dirname, 'styles')]
  }
}
```

`_document.js`:

```
import Document, { Head, Main, NextScript } from 'next/document'

export default class MyDocument extends Document {
  render() {
    return (
      <html>
        <Head>
          <link href="https://fonts.googleapis.com/css2?family=Lora:ital@0;1&family=Special+Elite&family=Noto+Sans+JP:wght@400;700&family=Reenie+Beanie&family=Raleway:wght@500&family=Oswald&family=Montserrat:wght@400;700&display=swap" rel="stylesheet" />
          <meta name="viewport" content="width=device-width, initial-scale=1" />
          <meta charSet="utf-8" />
        </Head>
        <body>
          <Main />
          <div id="dummy" />
          <NextScript />
        </body>
      </html>
    )
  }
}
```

Edit: When updating to 9.5.1 it gave me an error saying I shouldn't put meta tags in `_document.js` so they are since removed. Didn't help me fix the problem, though. Please don't tell me the problem is with my '@' setup? It's such a time saver!

Edit2: So I've discovered some more things. There is a `<style data-next-hide-fouc="true">body{display:none}</style>` inserted in the page source, right? I have a global css file that has `body` styling as well, and I insert it in my `_app.js`  with an `import '@/styles/global.scss'`.

When I remove the body styling, I get a white page instead of unstyled content. However, no css at all is actually preloaded.  What is the expected behavior, a white page for a long time, or some styled content (above-the-fold?) and then the rest following? I have a white page displaying for 4-5 seconds.

**Reply by aychtang** [reply]
@10xchs Would be interested to hear if you discovered more about the problem.

@Timer it would be a great help to have some update / assistance on this as we're also experiencing this issue intermittently in prod since our deployment with next.js.

**Comment by aychtang** [comment]
I'm also experiencing this with 9.4.x serverless target over aws lambda, will try upgrading to 9.5 now.

**Reply by aychtang** [reply]
It got a bit better, but you often still get strange FOUC. I don't have a particularly custom webpack config:

```
module.exports = {
  target: "serverless",
  webpack(config) {
    config.module.rules.push({
      test: /\.svg$/,
      issuer: {
        test: /\.(js|ts)x?$/,
      },
      use: ["@svgr/webpack"],
    });

    return config;
  },
  useFileSystemPublicRoutes: true,
};
```

**Reply by aychtang** [reply]
I don't really understand what is going on because when I look at the document all the CSS sheets are in the `head`. They all load far before DOMContentLoaded event, can someone help me understand 🤔 

<img width="1238" alt="Screenshot 2020-08-25 at 21 47 07" src="https://user-images.githubusercontent.com/1874468/91226158-900b9480-e71c-11ea-9204-b3aac25bc70e.png">
